From 8a2ecddf355f496f1c24b467de23ccee53a4f3a3 Mon Sep 17 00:00:00 2001
From: Subrahmanyam Nilla <snilla@marvell.com>
Date: Tue, 3 Mar 2020 16:00:40 +0530
Subject: [PATCH 05/19] octeontx2-dpi: Add MRRS and MPS configuration to dpi
 driver.

commit af0f961c704689dbb1a0a086abea10c8537845ce from
git@git.assembla.com:cavium/WindRiver.linux.git

Add module parameters to octeontx2-dpi driver to configure
Maximum Read Request Size(MRRS) and Maximum Payload size(MPS) during
driver load.

Change-Id: I67f818905698b9308869260f362e443ad6bc7ed6
Signed-off-by: Subrahmanyam Nilla <snilla@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/24368
Reviewed-by: Stanislaw Kardach <Stanislaw.Kardach@cavium.com>
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Signed-off-by: Abhijit Ayarekar <aayarekar@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/49028
Reviewed-by: Radha Chintakuntla <radhac@marvell.com>
Integrated-by: Abhishek Paliwal <paliwal.abhishek@windriver.com>
---
 drivers/soc/marvell/octeontx2-dpi/dpi.c | 47 +++++++++++++++++++++++--
 drivers/soc/marvell/octeontx2-dpi/dpi.h |  9 +++++
 2 files changed, 54 insertions(+), 2 deletions(-)

diff --git a/drivers/soc/marvell/octeontx2-dpi/dpi.c b/drivers/soc/marvell/octeontx2-dpi/dpi.c
index 9896a4488b40..ca95f37976a2 100644
--- a/drivers/soc/marvell/octeontx2-dpi/dpi.c
+++ b/drivers/soc/marvell/octeontx2-dpi/dpi.c
@@ -25,6 +25,15 @@ static const struct pci_device_id dpi_id_table[] = {
 	{ PCI_DEVICE(PCI_VENDOR_ID_CAVIUM, PCI_DEVID_OCTEONTX2_DPI_PF) },
 	{ 0, }  /* end of table */
 };
+
+static int mps = 128;
+module_param(mps, int, 0644);
+MODULE_PARM_DESC(mps, "Maximum payload size, Supported sizes are 128, 256, 512 and 1024 bytes");
+
+static int mrrs = 128;
+module_param(mrrs, int, 0644);
+MODULE_PARM_DESC(mrrs, "Maximum read request size, Supported sizes are 128, 256, 512 and 1024 bytes");
+
 MODULE_DEVICE_TABLE(pci, dpi_id_table);
 MODULE_AUTHOR("Marvell International Ltd.");
 MODULE_DESCRIPTION(DPI_DRV_STRING);
@@ -125,7 +134,8 @@ static int dpi_queue_fini(struct dpipf *dpi, struct dpipf_vf *dpivf, u8 vf)
  */
 static int dpi_init(struct dpipf *dpi)
 {
-	int engine = 0;
+	int engine = 0, port = 0;
+	u8 mrrs_val, mps_val;
 	u64 reg = 0ULL;
 
 	for (engine = 0; engine < dpi_dma_engine_get_num(); engine++) {
@@ -152,12 +162,39 @@ static int dpi_init(struct dpipf *dpi)
 	dpi_reg_write(dpi, DPI_DMA_CONTROL, reg);
 	dpi_reg_write(dpi, DPI_CTL, DPI_CTL_EN);
 
+	/* Configure MPS and MRRS for DPI */
+	if (mrrs < DPI_EBUS_MRRS_MIN || mrrs > DPI_EBUS_MRRS_MAX ||
+			!is_power_of_2(mrrs)) {
+		dev_info(&dpi->pdev->dev,
+			"Invalid MRRS size:%d, Using default size(128 bytes)\n"
+			, mrrs);
+		mrrs = 128;
+	}
+	mrrs_val = fls(mrrs) - 8;
+
+	if (mps < DPI_EBUS_MPS_MIN || mps > DPI_EBUS_MPS_MAX
+			|| !is_power_of_2(mps)) {
+		dev_info(&dpi->pdev->dev,
+			"Invalid MPS size:%d, Using default size(128 bytes)\n"
+			, mps);
+		mps = 128;
+	}
+	mps_val = fls(mps) - 8;
+
+	for (port = 0; port < DPI_EBUS_MAX_PORTS; port++) {
+		reg = dpi_reg_read(dpi, DPI_EBUS_PORTX_CFG(port));
+		reg &= ~(DPI_EBUS_PORTX_CFG_MRRS(0x7) |
+			 DPI_EBUS_PORTX_CFG_MPS(0x7));
+		reg |= (DPI_EBUS_PORTX_CFG_MPS(mps_val) |
+			DPI_EBUS_PORTX_CFG_MRRS(mrrs_val));
+		dpi_reg_write(dpi, DPI_EBUS_PORTX_CFG(port), reg);
+	}
 	return 0;
 }
 
 static int dpi_fini(struct dpipf *dpi)
 {
-	int engine = 0;
+	int engine = 0, port;
 	u64 reg = 0ULL;
 
 	for (engine = 0; engine < dpi_dma_engine_get_num(); engine++) {
@@ -170,6 +207,12 @@ static int dpi_fini(struct dpipf *dpi)
 	dpi_reg_write(dpi, DPI_DMA_CONTROL, reg);
 	dpi_reg_write(dpi, DPI_CTL, ~DPI_CTL_EN);
 
+	for (port = 0; port < DPI_EBUS_MAX_PORTS; port++) {
+		reg = dpi_reg_read(dpi, DPI_EBUS_PORTX_CFG(port));
+		reg &= ~DPI_EBUS_PORTX_CFG_MRRS(0x7);
+		reg &= ~DPI_EBUS_PORTX_CFG_MPS(0x7);
+		dpi_reg_write(dpi, DPI_EBUS_PORTX_CFG(port), reg);
+	}
 	return 0;
 }
 
diff --git a/drivers/soc/marvell/octeontx2-dpi/dpi.h b/drivers/soc/marvell/octeontx2-dpi/dpi.h
index 2d0cf04524b7..2026979d732a 100644
--- a/drivers/soc/marvell/octeontx2-dpi/dpi.h
+++ b/drivers/soc/marvell/octeontx2-dpi/dpi.h
@@ -184,6 +184,15 @@
 								((y) << 4))
 #define DPI_EPFX_PP_VF_LINTX_ENA_W1S(x, y)	(0x7800ULL | ((x) << 5) |\
 								((y) << 4))
+
+#define DPI_EBUS_MRRS_MIN			128
+#define DPI_EBUS_MRRS_MAX			1024
+#define DPI_EBUS_MPS_MIN			128
+#define DPI_EBUS_MPS_MAX			1024
+#define DPI_EBUS_MAX_PORTS			2
+#define DPI_EBUS_PORTX_CFG_MRRS(x)		(((x) & 0x7) << 0)
+#define DPI_EBUS_PORTX_CFG_MPS(x)		(((x) & 0x7) << 4)
+
 /* VF Registers: */
 #define DPI_VDMA_EN		(0x0ULL)
 #define DPI_VDMA_REQQ_CTL	(0x8ULL)
-- 
2.31.1

